samplers/parser.go

```diff
@@ -78,52 +69,127 @@ func ValidMetric(sample *UDPMetric) bool {
 	ret := true
 	ret = ret && sample.Name != ""
 	ret = ret && sample.Value != nil
-	if sample.SampleRate == 0 {
-		sample.SampleRate = 1
-	}
 	return ret
 }
 
-// ParseSSF takes in a byte slice and returns:
-// an SSFSpan, slice of UDPMetrics, and an error.
-// It also validates packets before returning them. Note that this function
-// is not currently safe for concurrent use. If needed in the future,
-// have the caller pass in a `proto.Buffer`!
-func ParseSSF(packet []byte) (*ssf.SSFSpan, []*UDPMetric, error) {
-	sample := &ssf.SSFSpan{}
-	scratchBuff.Reset()
-	scratchBuff.SetBuf(packet)
-	err := scratchBuff.Unmarshal(sample)
+// A Message struct represents a parsed SSF message. It encapsulates
+// an ssf.SSFSpan object, which can contain a trace span and/or a set
+// of metrics.
+type Message struct {
+	span *ssf.SSFSpan
+}
 
-	if err != nil {
-		return nil, nil, errors.New("unmarshal")
-	}
+type errInvalidTrace string
 
+// ErrInvalidTrace indicates that an SSF span was invalid.
+const ErrInvalidTrace errInvalidTrace = "span did not validate"
+
+func (e errInvalidTrace) Error() string {
+	return string(e)
+}
+
+// TraceSpan checks if an SSF message is a valid trace. If so, it
+// normalizes the span in-place (infers a Name from a "name" tag, sets
+// Tags if none were given), and returns that span.  If the span is
+// not a valid trace, TraceSpan returns a nil span and
+// ErrInvalidTrace.
+//
+// The span returned from TraceSpan does contain the (unparsed) metric
+// samples contained in the span.
+func (m *Message) TraceSpan() (*ssf.SSFSpan, error) {
+	if !ValidTrace(m.span) {
+		return nil, ErrInvalidTrace
+	}
+	return m.span, nil
+}
+
+// Metrics examines an SSF message, parses and returns any
+// metrics contained inside. If any parse error occurs in processing
+// any of the metrics, ExtractMetrics collects them into the error
+// type InvalidMetrics and returns this error alongside any valid
+// metrics that could be parsed.
+func (m *Message) Metrics() ([]*UDPMetric, error) {
```

Comment:
Is there a reason that we're returning `[]*UDPMetric` instead of `[]UDPMetric`? It's not bad per se, just slightly unusual (and indicative of desire to mutate), so I figured I'd ask :)

Reply:
That was the interface previously, think that's so they can be used for their MetricKey methods (which dispatch from a pointer)

Reply:
You can call `MetricKey` on it whether or not it's a pointer (the latter will be automatically referenced before calling the method). We can keep it for consistency though